WooCommerce/Classes/POS/Presentation/Card Present Payments/Connection Alerts/PointOfSaleCardPresentPaymentUpdatingReaderAlertViewModel.swift

```diff
@@ -0,0 +1,77 @@
+import Foundation
+import SwiftUI
+
+struct PointOfSaleCardPresentPaymentUpdatingReaderAlertViewModel {
+    let title: String
+    let image: Image
+    let progressTitle: String
+    let progressSubtitle: String?
+    let cancelButtonTitle: String
+    let cancelReaderUpdate: (() -> Void)?
+
+    init(requiredUpdate: Bool, progress: Float, cancel: (() -> Void)?) {
+        self.image = Image(uiImage: .softwareUpdateProgress(progress: CGFloat(progress)))
+
+        let isComplete = progress == 1
+
+        self.title = isComplete ? Localization.titleComplete : Localization.title
+        self.progressTitle = String(format: Localization.percentCompleteFormat, 100 * progress)
+        self.progressSubtitle = isComplete ? nil : (requiredUpdate ? Localization.messageRequired : Localization.messageOptional)
+
+        self.cancelButtonTitle = isComplete ? Localization.dismissButtonText :
+        (requiredUpdate ? Localization.cancelRequiredButtonText : Localization.cancelOptionalButtonText)
+        self.cancelReaderUpdate = cancel
```

Comment:
Could we have three different alerts here, to avoid optionals and make it more declarative? One for required updates, one for optional updates, one for complete updates?

Reply:
Yup makes sense, I'll push an update shortly. After checking the call sites in `CardReaderConnectionController` and  `BuiltInCardReaderConnectionController`, the `cancel` closure remains optional though in the in-progress views.

Reply:
Updated in https://github.com/woocommerce/woocommerce-ios/pull/13085/commits/9fb3cfbb3f8af24f225934f77db3556b4ec4aca6 with 3 views.